### PR TITLE
`cljr--reader-conditional-context` identifies current language

### PR DESCRIPTION
**Pull request by dgtized**
Addresses problems identified in
https://github.com/clojure-emacs/clj-refactor.el/issues/533.

Previously, `cljr--goto-reader-conditional` and`cljr--point-in-reader-conditional-branch-p` appeared to have problems with infinite loops as they looked for specific language values in the reader conditional.

`cljr--reader-conditional-context` starts at the beginning of the closest enclosing reader-conditional, and walks through each language, context pair returning the language name if the point is somewhere between the start of the language keyword or the next language pair. It will also bail and return nil if it searches beyond the scope of the reader-conditional or there is no language symbol it can identify.

`cljr--language-context-at-point` has been amended to use the `cljr--reader-conditional-context` to return both the language of the file & the current context to search from.

Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [x] You've added tests (if possible) to cover your change(s)
- [ ] The new code is not generating byte compile warnings (run `cask exec emacs -batch -Q -L . -eval "(progn (setq byte-compile-error-on-warn t) (batch-byte-compile))" clj-refactor.el`)
- [x] All tests are passing (run `./run-tests.sh`)
- [ ] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)

Thanks!

### CODE REVIEW

File: tests/unit-test.el [hunk 1]
```diff
@@ -197,57 +197,89 @@
...
+                :to-equal '("cljc" "cljs")))))
+
+  (it "returns nil in an incomplete reader conditional"
     (cljr--with-clojure-temp-file "foo.cljc"
       (with-point-at "(ns foo) #?(|)"
         (expect (cljr--language-context-at-point)
                 :to-equal '("cljc" nil)))))
 
-  (xit "returns :default context if specified"
+  (it "returns last context in an incomplete reader conditional"
```
[the remark this answers is not in the document]
Comment (dgtized) [reply tests/unit-test.el:242]: The first example was causing an unbounded loop, so I fixed it to stop if (point) reaches the end of the file. I think this was an edge case that only triggers if it's literally the last character in the file but regardless it's addressed. 3rd example worked fine.

However, the second example fails as `cljr--beginning-of-reader-conditional` fails to find the enclosing reader conditional, because `cljr--top-level-p` erroneously identifies the form as top level. I've added tests demonstrating this for `cljr--language-context-at-point`, `cljr--beginning-of-reader-conditional` and `cljr--top-level-p`. The result of an unparsable file is that we can't identify the context at point correctly (it identifies it as nil), but it does terminate, so that seems like an acceptable result given that the input is bad.

File: clj-refactor.el [hunk 2]
```diff
@@ -1947,6 +1947,43 @@ FEATURE is either :clj or :cljs."
...
                         "clj"))))))
 
+(defun cljr--beginning-of-reader-conditional ()
+  "Return position of beginning of containing reader conditional.
+
+Otherwise `nil' if outside of reader conditional."
+  (save-excursion
+    (let ((reader-start nil))
+      (while (not (or reader-start (cljr--top-level-p)))
+        (backward-up-list)
```
[the remark this answers is not in the document]
Comment (dgtized) [reply clj-refactor.el:1957]: I would prefer if we used more the builtin functions instead of depending on paredit. In this case it depends on `cljr--depth-at-point`, which depends on `paredit-current-parse-state` (which is a soft wrapper around `beginning-of-defun` and `parse-partial-sexp` both of which are builtin) and `paredit-in-string-p`, which also depends on parse state from `parse-partial-sexp`. As discussed below, it appears that `cljr--top-level-p` doesn't handle files with preceeding unmatched parentheses, if we need to address this, I think we would need to dig into the `syntax-ppss` functions emacs provides in `syntax.el`.

If there are specific edge cases that paredit-backwards-up handles over backward-up-list then that would make sense to stay with paredit, but I would prefer to use the out of the box sexp navigation if it suffices. I think it makes a lot of sense to use paredit when manipulating sexps as it adds a lot to ensure the file remains parseable, I'm just not aware of the advantage for simple sexp navigation.

**Comment by dgtized** [comment]
I think it would be good to do a cleanup PR that deprecates and removes any of the existing language context path to use the new approach here. I would rather just attach that to #533 to close it down, but also happy to make a new issue.

Separately though, I think we should start thinking about when to enable the `cljr-slash-uses-suggest-libspec` by default and start the process of deprecating the old way. I think almost everything for that path is done on the emacs side, we can definitely continue to improve in the middleware, but it would be useful to start nudging users in that direction. I think the two big issues remaining from the original scope were both to add language context support to the middleware, and to get the middleware to parse the user configured namespaces that incorporates language context information. Both of which are additional features ontop of what we had before the feature flag was introduced. Anyway, let me know if we should start that process.
